File: common/src/main/java/dev/qixils/aviary/common/messaging/serializers/impl/BooleanSerializer.java
package dev.qixils.aviary.common.messaging.serializers.impl;

import dev.qixils.aviary.common.messaging.serializers.SingletonByteSerializer;
import org.jetbrains.annotations.NotNull;

/**
 * A {@link SingletonByteSerializer} for {@link Boolean}s.
 */
public class BooleanSerializer implements SingletonByteSerializer<Boolean> {
	private static final byte TRUE = (byte) 0x01;
	private static final byte FALSE = (byte) 0x00;


	@Override
	public byte @NotNull [] serialize(@NotNull Boolean aBoolean) {
		return new byte[] { aBoolean ? TRUE : FALSE };
	}

	@Override
	public @NotNull Boolean deserialize(byte @NotNull [] data) {
		if (data.length != 1)
			throw new IllegalArgumentException("data must be of length 1");
		return deserialize(data[0]);
	}

	@Override
	public @NotNull Boolean deserialize(byte data) {
		if (data == TRUE)
			return true;
		else if (data == FALSE)
			return false;
		else
			throw new IllegalArgumentException("data must be either 0x00 or 0x01, not " + data);
	}
}

File: common/src/main/kotlin/dev/qixils/aviary/common/messaging/serializers/impl/BooleanSerializer.kt
package dev.qixils.aviary.common.messaging.serializers.impl

import dev.qixils.aviary.common.messaging.serializers.SingletonByteSerializer

/**
 * A [SingletonByteSerializer] for [Boolean]s.
 */
class BooleanSerializer : SingletonByteSerializer<Boolean> {

	override fun serialize(obj: Boolean): ByteArray {
		return byteArrayOf(if (obj) TRUE else FALSE)
	}

	override fun deserialize(data: ByteArray): Boolean {
		require(data.size == 1) { "data must be of length 1" }
		return deserialize(data[0])
	}

	override fun deserialize(data: Byte): Boolean {
		return when (data) {
			TRUE -> true
			FALSE -> false
			else -> throw IllegalArgumentException("data must be either 0x00 or 0x01, not 0x%02x".format(data))
		}
	}

	companion object {
		private const val TRUE = 0x01.toByte()
		private const val FALSE = 0x00.toByte()
	}
}
